Enable publishing with local dependencies and workspace support

markusz opened the pull request:
## Summary

  This PR adds support for local dependencies, workspace management, and refactors the codebase for better maintainability.

  ## Problems Solved

  - **No recursive local dependency installation** - Users had to manually `cd` into each local dependency and run `buffrs install`
  - **Publishing with local dependencies failed** - Packages couldn't reference local dependencies when publishing
  - **No workspace support** - Multi-package projects required manual directory navigation and order-dependent operations
  - **Order-sensitive operations** - Dependency processing order affected behavior

  ## Key Changes

  ### Workspace Support
  - `buffrs install` - Processes all workspace members and their dependencies
  - `buffrs publish` - Publishes workspace members with automatic dependency ordering and deduplication
  - `buffrs uninstall` - Clears vendor directories for all workspace members
  - Incompatible commands provide clear error messages directing users to package directories

  ### Complete modular rewrite of Dependency Resolver (`src/resolver.rs`)
  - **Separation of concerns**: Dependency resolution logic extracted into dedicated, testable module
  - **Reusability**: Used by both `install` and `publish` commands
  - **Topological sorting**: Ensures dependencies are processed before dependents
  - **Clear error handling**: Detects circular dependencies and validates package type constraints

  ### Publishing with Local Dependencies
  - Local dependencies are published first in topological order
  - Published packages have local references replaced with remote registry references
  - Idempotent publishing prevents duplicate uploads

  ### Code Architecture
  - Extracted `Installer` helper (`src/operations/installer.rs`) - reduced `install()` from ~200 to 4 lines
  - Refactored `Publisher` helper (`src/operations/publisher.rs`) - reduced `publish()` from ~140 to 20 lines
  - Commands now act as thin wrappers delegating to specialized helpers

  ## Testing
  - Added 26 unit tests for existing but untested core functionality
  - Added 19 unit tests dedicated to new dependency resolver
  - All 102 (from 46) tests passing (65 unit + 17 e2e + 19 resolver + 1 doc)

  ## Example

  **Before:**
  ```bash
  cd local-lib && buffrs publish
  cd ../package-a && buffrs publish  # Fails - can't reference local deps

  After:
  cd workspace-root && buffrs publish
  # Automatically publishes local-lib, then package-a with registry references
  ```
  ## Breaking Changes

  None. All existing functionality is preserved.

Commit:
feat(cli): add basic (de)serialization support for workspace section in Proto.toml

Commit:
feat(cli): Increase minor version of CLI

Commit:
feat(cli): Make dependencies optional to account for new workspace config

Commit:
feat(cli): Check for mutual exclusivity between workspace and dependencies

- Proto.toml can either contain a workspace or dependencies, but not both
- Make dependency Option
- Add tests for invalid manifest configurations

Commit:
chore(cli): Remove misleading TODO

Commit:
feat(cli): Remove check for PackageType::Api in resolve()

- Remove the check that prevents packages with type = api to be used as dependencies. The check prevents effective use of the workspace feature and also contradicts docs & reality of existing projects
- Expand documentation of pub API of store.rs

Commit:
feat(cli): Add ManifestType enum to Manifest

- Add enum to express type of manifest
- Add checks and error handling for invalid manifest configurations
- Manifest::new may Err now -> return Result<>

Commit:
feat(cli): Upgrade crate to v0.12.0

- Reflect changes to Cargo.toml
- Adapt fixtures for integration test & document footguns
- Bump versions of dependencies toml & gix to latest minor
- Add contributor to authors

Commit:
feat(cli): Add support to run install for workspaces

- Parse and apply glob rules in members and exclude attribute
- Make processing path aware for Manifest, Lockfile, PackageStore and DependencyGraph to allow for running command::install in subfolders
- Add glob and tempfile dependency

markusz committed:
Merge pull request #1 from markusz/mz/feat/add-workspace-kw-to-toml

Add support for workspace keyword in Proto.toml files

markusz committed:
Merge pull request #2 from markusz/mz/feat/add-workspace-install

Add CLI support to run buffrs install on a workspace

mara-schulke commented on src/manifest.rs:
```diff
@@ -284,17 +416,47 @@ pub struct Manifest {
 
 impl Manifest {
+    /// Determine the ManifestType based on dependencies and workspace
+    fn get_manifest_type(
```
This indicates that you'd want to have a `enum Manifest` instead of `struct Manifest` with a bunch of internal options!

markusz replied:
In an earlier version of the code, I had something like this.

```
pub struct PackageManifest {
    ..
}

pub struct WorkspaceManifest {
    ..
}


pub enum BuffrsManifest {
    Workspace(WorkspaceManifest),
    Package(PackageManifest)
}

impl TryFrom<RawManifest> for BuffrsManifest {
    type Error = ();

    fn try_from(value: RawManifest) -> Result<Self, Self::Error> {
        todo!()
    }
}
```

I discarded it, since it would be quite invasive and I wasn't sure I understood the codebase + changes required enough to justify such a large change.

@mara-schulke Is that the direction in which you are thinking?

mara-schulke replied:
Yes exactly; I am absolutely for a larger change in this regard if that implies that the whole `ManifestType` thing is not needed anymore

mara-schulke commented on src/command.rs:
```diff
@@ -70,7 +72,7 @@ pub async fn init(kind: Option<PackageType>, name: Option<PackageName>) -> miett
         .transpose()?;
 
-    let manifest = Manifest::new(package, vec![]);
+    let manifest = Manifest::new(package, Some(vec![]), None)?;
```
Optional arguments are usually suboptimal in signatures; I'd recommend using a build pattern:

```rust
let manifest = Manifest::builder().package(pkg).dependencies(vec![]).build();
```

The typed builder crate is great for inspo; Given that `Manifest` should likely be an enum, i'd recommend writing your own builder.

markusz replied:
Agreed, let's go with a builder pattern here. There are uses of Optional args in the existing codebase (e.g. [here](https://github.com/helsing-ai/buffrs/blob/main/src/manifest.rs#L292)) which I'll convert to builder pattern where appropriate

mara-schulke commented on src/command.rs:
```diff
@@ -360,13 +367,72 @@ pub async fn publish(
+}
+
+/// Installs dependencies for a workspace (not yet implemented)
+pub async fn install_workspace(preserve_mtime: bool, manifest: &Manifest) -> miette::Result<()> {
```
Is this its own command and useful to have in the crates public api? `pub async fn` in this mod are meant to map 1:1 to `buffrs <command>`

markusz replied:
No it's not. The type of the manifest should be transparent to the user running `buffrs install`. Following the convention, both `install_package` and `install_workspace` should be private

mara-schulke commented on src/command.rs:
```diff
@@ -384,6 +450,7 @@ pub async fn install(preserve_mtime: bool) -> miette::Result<()> {
         &credentials.into(),
         &cache,
         preserve_mtime,
+        Some(package_path),
```
Another instance of optional arguments;

Also why is the pkg path needed here?

markusz replied:
package path is required because with workspace support for `buffrs install`, the assumption that the buffrs process runs in the root of a buffrs project does not hold true anymore

E.g.

```
cd ~/project_a # a buffers workspace
buffrs install # current_dir of process is ~/project_a
```

buffrs will now find the workspace members and run `install_package` on those. We can't rely on the existing path logic and need to provide the appropriate relative path (seen from the project root)

markusz replied:
Good catch, package_path doesn't need to be optional, as we always have a path that we work on, even if it's just the project root

mara-schulke commented on src/manifest.rs:
```diff
@@ -275,6 +302,111 @@ impl FromStr for RawManifest {
+///
+/// https://doc.rust-lang.org/cargo/reference/workspaces.html#the-members-and-exclude-fields
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Workspace {
```
I think this could go into its own `workspace.rs`? As it also holds logic beyond just defining how workspace is serialized in the manifest

markusz replied:
Agreed, workspace grew beyond just being a struct and deserves it's own mod

mara-schulke commented on src/manifest.rs:
```diff
@@ -307,17 +469,20 @@ impl Manifest {
     pub async fn read() -> miette::Result<Self> {
-        Self::try_read_from(MANIFEST_FILE)
-            .await?
-            .ok_or(miette!("`{MANIFEST_FILE}` does not exist"))
```
Why did you remove the miette trace?

markusz replied:
Collateral damage of the refactoring. I'll re-add it

Commit:
refactor(cli): Move workspace to dedicated module

- Fix for https://github.com/markusz/buffrs/pull/3#discussion_r2419047381

Commit:
refactor(cli): Make install helpers private and improve documentation

- Fix for https://github.com/markusz/buffrs/pull/3#discussion_r2419032024

Commit:
fix(cli): Readds trace for manifest::read()

- Fix for https://github.com/markusz/buffrs/pull/3#discussion_r2419067149

Commit:
refactor(cli): package_path can be non-Option as a path is always required when processing a dependency

- Fix for https://github.com/markusz/buffrs/pull/3#discussion_r2419038672

Commit:
refactor(cli): Introduce builder pattern for Manifest

- Fix for https://github.com/markusz/buffrs/pull/3#discussion_r2419020631

Commit:
ci(cli): Add more units tests to manifest.rs

Commit:
fix(cli): Make lockfile handling in dependency resolver subpath aware

Commit:
fix(cli): Re-add check that packages with type = lib can not depend on packages of type = api

- Evaluate validity of dependencies while build DependencyGraph
- Add tests for valid and invalid behaviour

Commit:
feat(cli): Split buffrs publish fn in publish_workspace & publish_package

Commit:
refactor(cli): Implement new version of resolver

Goals:

- Simplify code
- Separate concerns of building, validating, topo sorting and operating (process, install) on nodes
- Add missing cycle detection
- Enable in-order processing through topological sorting

Commit:
test(cli): Add extensive tests to graph building algorithm.

Tests cover:
- graph construction
- validation
- cycle detection
- error cases
- different topologies (diamond dep, ..)

Commit:
feat(cli): Add topological sort on DependencyGraph

- Implement Kahn's algorithm to convert DependencyGraph into dependency-ordered list of PackageName
- Add tests for various graphs incl. edge cases

Commit:
refactor(cli): Reduce file noise by moving tests from resolver_v2.rs into dedicated test file

Commit:
feat(cli): [WIP] Add base logic for install via v2 of resolver

Commit:
feat(cli): [WIP] Minor formatting

Commit:
docs(cli): Add documentation to existing functionality

Commit:
docs(cli): Add documentation to existing functionality

Commit:
fix(cli): Add check for version mismatch of two remote dependencies

Commit:
fix(cli): Add cache logic to install_package

Commit:
feat(cli): Make v2 of the resolver the default implementation

- Separates concerns: Building a dependency tree, rule based validation, and topological sorting are now separate
- Mutating CLI operations (like install or publish) are not required to happen in the recursive function parsing the dependency tree. Instead, they operate in-order on an ordered Vector of dependencies retrieved by DependencyGraph
- Separation of concerns allows for better unit testability of the code, relying less on finicky e2e tests

Commit:
style(cli): Add compliance with all fmt and clippy rule sets

Commit:
chore(cli): Code hygene

Commit:
feat(cli): Add first [WIP] version of buffrs publish on workspaces

Commit:
feat(cli): [WIP] Add publishing logic.

- Conceptually, impl is complete, but various e2e tests fail

Commit:
fix(cli): Publish root package after dependencies have been published

Commit:
fix(cli): Adapt e2e tests to account for changed behavior of publishing

- publish should publish local deps to remote and make them available to dependent packages

Commit:
fix(cli): Fix predicate error in get_dependencies_of_type

Commit:
fix(cli): Fix test fixtures for local publishing test

Commit:
fix(cli): Fix error related to absolute vs relative paths being used as keys in the HashMap

Commit:
chore(cli): Cleanup of debugging statements

Commit:
refactor(cli): Add Publisher helper to encapsulate publishing logic

Extracted publishing operations into a dedicated Publisher struct to reduce parameter passing and improve code organization.

Commit:
refactor(cli): Add Installer helper to encapsulate publishing logic

Extracted installing operations into a dedicated Installer struct to reduce parameter passing and improve code organization.

Commit:
test(cli): Add unit tests for core functionality of Publisher and Installer helper

Commit:
chore(cli): Housekeeping, formatting, linting

Commit:
fear(cli): Implement workspace publish and moves behind clean API in publisher

Commit:
fear(cli): Add ensure_package_manifest() helper for commands that can only be ran on packages

Commit:
fear(cli): Add support for uninstalling a workspace

Commit:
chore(cli): Remove references of resolver v2 as this becomes the default in v0.12

markusz committed:
Merge pull request #4 from markusz/mz/add-workspace-publish

Draft: Add support for buffrs publish on [workspace]

Commit:
chore(cli): Housekeeping

Commit:
docs(cli): Incorporate changes buffrs book docs

Commit:
chore(cli): Implements miette best practices and fixes other minor nits

Commit:
refactor(cli): Extracts validation check into fn

asmello commented on src/lib.rs:
```diff
@@ -33,13 +33,22 @@ pub mod lock;
+/// Dedicated helpers to deal with more complicated CLI operations
+pub mod operations {
+    /// Installer helper to handle logic related to installing dependencies
+    pub mod installer;
```
**note**: I genuinely didn't know this kind of nested declarations were a thing!

markusz replied:
Yeah its not used that often, but was the recommended standard at my previous job. Muscle memory I guess..

I'll change it to be uniform with the patterns used in the existing codebase

asmello commented on src/lock.rs:
```diff
@@ -12,11 +12,11 @@
 use semver::Version;
 use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use std::path::Path;
```
**nit**: I don't know if `buffrs` has a formal policy on imports, but seems like the others have been merged so I suggest doing the same here.

mara-schulke replied:
I am in general for:

1. Merging imports when sensible
2. Using the `std ext crate` ordering (which used to be the case before this mr)

asmello commented on src/lock.rs:
```diff
@@ -191,8 +206,10 @@ impl Lockfile {
             packages,
         };
 
+        let lockfile_path = path.as_ref().to_path_buf().join(LOCKFILE);
```
**suggestion**: avoid double allocation by removing `.to_path_buf()`, it's not needed.

asmello commented on src/lock.rs:
```diff
@@ -172,8 +178,17 @@ impl Lockfile {
         }
     }
 
+    /// Loads the Lockfile from a specific path, if it exists, otherwise returns an empty one
```
**suggestion**: worth documenting what happens when checking if the file exists fails (seems that it correctly throws an error, rather than coercing to false).

asmello commented on src/command.rs:
```diff
@@ -277,193 +316,79 @@ pub async fn publish(
+                let canonical_name = fs::canonicalize(&package_path).await.into_diagnostic()?;
+                tracing::info!(
+                    ":: uninstalling dependencies for package: {}",
+                    canonical_name.display()
```
**thought**: for display purposes, maybe the non-canonical name is more friendly?

markusz replied:
Fixed in https://github.com/markusz/buffrs/pull/3/commits/5a10284fb1677841be924d9628a78d0f7f3aa490

asmello commented on src/command.rs:
```diff
@@ -232,15 +271,17 @@ pub async fn package(
-    if let Some(version) = version {
-        if let Some(ref mut package) = manifest.package {
-            tracing::info!(":: modified version in published manifest to {version}");
+    if let Some(version) = version
+        && let Some(ref mut package) = manifest.package
```
**praise**: very nice! Though this does raise the minimum compiler version. Maybe we should set it explicitly in the manifest?

asmello commented on src/command.rs:
```diff
@@ -199,26 +231,33 @@ pub async fn add(registry: RegistryUri, dependency: &str) -> miette::Result<()>
 
     manifest
         .dependencies
+        .get_or_insert_with(Vec::new)
```
```suggestion
        .get_or_default()
```

asmello commented on src/command.rs:
```diff
@@ -104,12 +130,16 @@ pub async fn new(kind: Option<PackageType>, name: PackageName) -> miette::Result
+    if let Some(pkg) = package {
+        builder = builder.package(pkg);
+    }
+    let manifest = builder.dependencies(vec![]).build();
```
**suggestion**: I think this could just be `builder.build();`. Same above.

asmello commented on src/command.rs:
````diff
@@ -38,6 +37,32 @@ use tokio::{
+/// }
+/// ```
+async fn ensure_package_manifest(command_name: &str) -> miette::Result<()> {
+    let manifest = Manifest::read().await?;
````
**thought**: I don't think this is a performance bottleneck at all, but it does seem a bit unfortunate to parse the file once, just for this check, and discard the result. Maybe you could return the parsed manifest from here and reuse it? Maybe a TODO/future improvement though.

asmello commented on src/command.rs:
```diff
@@ -13,16 +13,15 @@
 
-use async_recursion::async_recursion;
+use crate::operations::installer::Installer;
+use crate::operations::publisher::Publisher;
```
**nit**: please merge

asmello commented on src/registry/cache.rs:
```diff
@@ -115,15 +115,15 @@ mod tests {
-            vec![],
-        );
+            })
+            .dependencies(vec![])
```
**suggestion**: remove

mara-schulke replied:
Yep, this should be populated using `Default::default()`

asmello commented on tests/cmd/add/mod.rs:
```diff
@@ -1,5 +1,9 @@
+/// This is a finicky test that depends on the contents of lib-0.0.1.tgz. It will fail if the crate version changes.
+///
+/// If you need to change the version, do the following:
+/// PACKAGE_VERSION=0.12 && mkdir tmp && tar -xzf lib-0.0.1.tgz -C tmp && sed -i '' "s/^edition = .*/edition = \"$PACKAGE_VERSION\"/" tmp/Proto.toml && COPYFILE_DISABLE=1 tar -czf lib-0.0.1.tgz -C tmp . && rm -rf tmp
```
**suggestion**: add ``` around this command

asmello commented on src/workspace.rs:
```diff
@@ -0,0 +1,101 @@
+        use crate::manifest::MANIFEST_FILE;
+        use miette::{IntoDiagnostic, WrapErr};
+        use std::collections::BTreeSet;
+        use std::fs;
```
**nit**: pls merge.

**question**: any particular reason to not have these at the top level?

markusz replied:
No good reason, likely IntelliJ being drunk

Fixed in https://github.com/markusz/buffrs/pull/3/commits/fe1521006d57b84a8e7018a9e0a7f196b88e7353

asmello commented on src/workspace.rs:
```diff
@@ -0,0 +1,101 @@
+
+        // Default to ["*"] if members is not specified
+        let default_members = vec!["*".to_string()];
+        let member_patterns = self.members.as_ref().unwrap_or(&default_members);
```
**nit**: `.unwrap_or_else(|| vec!["*".to_string()]);`

Absolutely doesn't make a practical difference, but it's good to maintain good habits.

asmello commented on src/workspace.rs:
```diff
@@ -0,0 +1,101 @@
+
+        // Process each member pattern
+        for pattern in member_patterns {
+            if pattern.contains('*') || pattern.contains('?') || pattern.contains('[') {
```
**suggestion**: use `pattern.contains(['*', '?', '['])` (I _think_ clippy can make this suggestion).

asmello commented on src/workspace.rs:
```diff
@@ -0,0 +1,101 @@
+                    if path.is_dir()
+                        && let Some(name) = path.file_name().and_then(|n| n.to_str())
+                        && pattern_matcher.matches(name)
+                        && path.join(MANIFEST_FILE).exists()
```
**suggestion**: I think it's worth reporting an error using `try_exists()` instead of silently ignoring members for which the call fails.

asmello commented on src/workspace.rs:
```diff
@@ -0,0 +1,101 @@
+            } else {
+                // Literal path - check if it exists and has Proto.toml
+                let member_path = workspace_root.as_ref().join(pattern);
+                if member_path.is_dir() && member_path.join(MANIFEST_FILE).exists() {
```
**suggestion**: similar to above, `.exists()` is generally discouraged.

asmello commented on src/workspace.rs:
```diff
@@ -0,0 +1,101 @@
+        let final_members: Vec<PathBuf> = resolved_members
+            .into_iter()
+            .filter(|member| {
+                let member_str = member.to_str().unwrap_or("");
```
```suggestion
                let member_str = member.to_str()?;
```

asmello commented on src/workspace.rs:
```diff
@@ -0,0 +1,101 @@
+
+                // Check if this member matches any exclude pattern
+                !exclude_patterns.iter().any(|exclude_pattern| {
+                    if let Ok(glob_matcher) = glob::Pattern::new(exclude_pattern) {
```
**question**: what does it mean when this fails? Could you just apply the same implicit check above instead of manually looking for `*`, `?` or `[`?

markusz replied:
The if case tests if the pattern can be treated as a glob pattern. If so, it tries to match the member (=path as str). If its not a glob pattern it's considered an exact path that is just compared.

This can be written more computationally efficient by precomputing the valid exclude glob patterns. However, as this is not a performance critical path, I optimized for readability over performance

asmello commented on src/operations/installer.rs:
```diff
@@ -0,0 +1,470 @@
+use std::path::{Path, PathBuf};
+use tokio::fs;
+
+use crate::package::PackageName;
```
**nit**: please merge imports to match crate convention

mara-schulke replied:
This, and std-ext-crate ordering

asmello commented on src/operations/installer.rs:
```diff
@@ -0,0 +1,470 @@
+// Copyright 2023 Helsing GmbH
```
**note**: since this is a new file, I think it should be 2025?

asmello commented on src/operations/installer.rs:
**note**: didn't look closely at this file, since it looks like a refactor of existing code. Please flag if you've made any non-trivial changes to the algorithm.

asmello commented on src/operations/publisher.rs:
```diff
@@ -0,0 +1,587 @@
+use std::env;
+use std::path::Path;
+use std::str::FromStr;
+use tokio::fs;
```
**nit**: merge imports

asmello commented on src/operations/publisher.rs:
```diff
@@ -0,0 +1,587 @@
+            .await;
+
+        let Ok(output) = output else {
+            return Ok(Vec::new());
```
**question**: isn't this really an error case? Same for the next check below.

asmello commented on src/operations/publisher.rs:
**thought**: I don't know if we do TODOs here, but feels like it'd be worth documenting [using proper `git` bindings](https://github.com/rust-lang/git2-rs) as a future enhancement. Maybe even [`gitoxide`](https://github.com/GitoxideLabs/gitoxide).

mara-schulke replied:
We had this before, introducing libgit is a massive pain as it introduces a chain of deps. I agree the current impl is rather brittle, but better than pulling in the whole openssl dep chain.

asmello commented on src/operations/publisher.rs:
```diff
@@ -0,0 +1,587 @@
+
+    /// Gets the list of files with uncommitted changes from git
+    #[cfg(feature = "git")]
+    async fn get_git_statuses() -> miette::Result<Vec<String>> {
```
**thought**: this name seems misleading to me, isn't this more of a `get_changed_files`?

asmello commented on src/operations/publisher.rs:
```diff
@@ -0,0 +1,587 @@
+    async fn publish_workspace_from_manifest(
+        &mut self,
+        manifest: &Manifest,
+        _package_path: &Path,
```
**question**: why is this unused?

markusz replied:
Fixed in 2f67e19b6ec77dc8725f3a6119d3e5cee3e4ff81

asmello commented on src/operations/publisher.rs:
```diff
@@ -0,0 +1,587 @@
+        })?;
+
+        if version.is_some() {
+            tracing::warn!(":: version flag is ignored for workspace publishes");
```
**thought**: maybe this should be a hard error (and flagged earlier)?

asmello commented on src/operations/publisher.rs:
```diff
@@ -0,0 +1,587 @@
+            let canonical_name = fs::canonicalize(&member_path).await.into_diagnostic()?;
+            tracing::info!(
+                ":: processing workspace member: {}",
+                canonical_name.display()
```
**note**: as elsewhere, maybe this is better off as a non-absolute path?

asmello reviewed (commented):
General point, would have been better to do the refactoring separately, to facilitate review.

I haven't reviewed `manifest.rs`, `resolver.rs` and `resolver_tests.rs` yet, but I need to pivot. Will come back when I have a chance. Generally looks pretty good though!

mara-schulke commented on docs/src/guide/workspaces.md:
```diff
@@ -0,0 +1,122 @@
+members = [
+  "packages/common",
+  "packages/api-v1",
+  "packages/api-v2",
```
Nit: This would (traditionally be a non goal for buffrs)

mara-schulke replied:
Like, having `-v1` and `-v2` suffixes to buffrs packages partly defeats it's purpose. Maybe a different example makes would be less self contradicting for buffrs?

Could just be `api-one` and `api-two` without any versioning

markusz replied:
Fixed in 31f40ef6cc84c5dcdfb8f3cf0dfcec5a233d8847

mara-schulke commented on docs/src/guide/workspaces.md:
```diff
@@ -0,0 +1,122 @@
+
+When publishing the workspace, buffrs automatically publishes `my-common` first, then publishes `my-api-v2` with the dependency reference updated to point to the registry.
+
+## Workspace-only Commands
```
```suggestion
## Package-only Commands
```

mara-schulke replied:
You are writing that the following list of comments is just applicable to packages, not to workspaces?

mara-schulke commented on src/manifest.rs:
```diff
@@ -284,19 +319,148 @@ pub struct Manifest {
+            package: self.package,
+            dependencies: None,
+            workspace: Some(self.state.workspace),
+            manifest_type: ManifestType::Workspace,
```
Phantom comment for the resolved manifest type thread; I think we should still implement this.

markusz replied:
I'm not very fond of that enum either, so happy to rework this to a nicer solution.

[Previous comment](https://github.com/markusz/buffrs/pull/3#discussion_r2419016739) got buried. Can you have a look if this is what you have in mind.

mara-schulke commented on src/package/store.rs:
```diff
@@ -111,7 +160,19 @@ impl PackageStore {
+    ///
+    /// Removes the package directory `proto/vendor/<package_name>/` and all its contents.
+    ///
+    /// # Arguments
```
Generally i'd vouch agains this as code should be self documenting for the most part and rust doc is great at documenting arguments; Keeping the error cases seems sane, although ideally we'd use a `thiserror` type, which is in turn: self documenting

markusz replied:
Reduced doc noise for most fn's in 9fb1aa1bb1adee7b79070951add0e9ab8d3ff6ae.

Kept more extensive documentation for some of the more involved functions (populate, release, ..) that I struggled to grok

Commit:
refactor(cli): Unifies module declaration

Commit:
refactor(cli): Apply std ext crate and nested imports across codebase

- Fixes https://github.com/markusz/buffrs/pull/3#discussion_r2444158496

Commit:
refactor(cli): Fix minor nits and suggestions

Adresses:

- https://github.com/markusz/buffrs/pull/3#discussion_r2444163962
- https://github.com/markusz/buffrs/pull/3#discussion_r2444180417
- https://github.com/markusz/buffrs/pull/3#discussion_r2444186422

Commit:
refactor(cli): return manifest in package manifest check

- Fixes https://github.com/markusz/buffrs/pull/3#discussion_r2444195294

Commit:
refactor(cli): Fixes various nits from review

Commit:
refactor(cli): Fixes various nits from review

Commit:
refactor(cli): Exit if workspace path is invalid UTF8

Commit:
refactor(cli): Fix imports

Commit:
refactor(cli): add copyright header to new files according to Helsing guidelines

Commit:
refactor(cli): nest use statements in resolver_tests.rs

Commit:
refactor(cli): Change naming of git check to better name

Commit:
refactor(cli): Remove unused arg from publish_workspace_from_manifest signature

Commit:
refactor(cli): Change example and error in docs

Commit:
refactor(cli): Reduce doc noise in store.rs

Commit:
refactor(cli): Simplify signature of publish_package_from_manifest by moving version one level up

Commit:
refactor(cli): Add first 90% working version of Manifest v2

Commit:
refactor(cli): Resolve all build errors, adapt code to new manifest representation

Commit:
fix(cli): Fix all unit tests

Commit:
refactor(cli): Merge logic + tests from temporary manifest_v2.rs back to manifest.rs

Commit:
chore(cli): Apply clippy rules to import ordering

Commit:
fix(cli): Add correct edition handling in manifest serialization

Commit:
fix(cli): Fix latent bug in read_from_or_default + add unit tests

- Fixes https://github.com/markusz/buffrs/pull/3/files/756db0740c585187fb20e8f76d668195d5cc3319#diff-9ce4c5920c454f1b2d1d0f7f48d7341aa91960831d15cfe53ce0a05dc6c1eeb5

Commit:
docs(cli): Documents read_or_default  behaviour in lock.rs impl

Commit:
fix(cli): Bail instead of WARN log for use of version flag during workspace publish

- Fixes: https://github.com/markusz/buffrs/pull/3#discussion_r2444693361

Commit:
Merge branch 'mz/feat/add-workspace-support' into mz/add-workspace-support/redesign-manifest-impl

Commit:
style(cli): Apply std ext crate & nested deps rules to imports

Commit:
style(cli): Nest tokio imports in command.rs

Commit:
style(cli): Formatting

Commit:
test(cli): Add extensive unit tests to manifest.rs

mara-schulke commented on src/manifest.rs:
```diff
@@ -275,6 +303,14 @@ impl FromStr for RawManifest {
+    Package,
+    /// The Manifest defines a workspace
+    Workspace,
+}
```
```suggestion
}

```

Nit: Missing newline

mara-schulke commented on src/manifest.rs:
```diff
@@ -284,19 +320,148 @@ pub struct Manifest {
-    /// Create a new manifest of the current edition
-    pub fn new(package: Option<PackageManifest>, dependencies: Vec<Dependency>) -> Self {
+    /// Clones the Manifest but replaces the dependencies with a given Vec
+    pub fn clone_with_different_dependencies(&self, dependencies: Vec<Dependency>) -> Self {
```
```suggestion
    pub fn with_dependencies(&self, dependencies: Vec<Dependency>) -> Self {
```

For clarity?

mara-schulke commented on src/manifest.rs:
```diff
@@ -518,3 +699,461 @@ impl From<LocalDependencyManifest> for DependencyManifest {
+            assert_eq!(cloned_raw_manifest_str, raw_manifest_str);
+        }
+    }
+    mod manifest_tests {
```
```suggestion

    mod manifest_tests {
```

Nit: Missing newline

mara-schulke commented on src/manifest.rs:
```diff
@@ -518,3 +699,461 @@ impl From<LocalDependencyManifest> for DependencyManifest {
+            assert_eq!(members[2], PathBuf::from("zebra"));
+        }
+    }
+}
```
Overall this looks sensibel but the manifest type bit is still unaddressed. I am not opposed to merging this (mostly) as is, but it is unclear to me whether my comments have been addressed or were resolved; either way this looks good:)

mara-schulke commented on src/resolver.rs:
```diff
@@ -1,469 +1,478 @@
-                .map(|sub_dependency| sub_dependency.package.clone())
-                .collect();
+        // Validate package type constraint
+        if let (Some(PackageType::Lib), Some(PackageType::Api)) = (parent_type, package_type) {
```
Maybe use 

```suggestion
        if let Some(PackageType::Lib) = parent_type && Some(PackageType::Api) = package_type {
```

?

mara-schulke commented on src/resolver.rs:
```diff
@@ -1,469 +1,478 @@
+        let version = &remote_manifest.version;
+
+        // Download the package to read its manifest and discover dependencies
+        let artifactory = Artifactory::new(registry.clone(), self.credentials)
```
Please note that this is about to change as maven support is actively being implemented. Is there a straight forward way to migrate this bit?

mara-schulke commented on src/resolver.rs:
```diff
@@ -1,469 +1,478 @@
+            (DependencyManifest::Local(_), DependencySource::Remote { .. }) => {
+                bail!(DependencyError::LocalRemoteConflict {
+                    local_pkg: dependency.package.clone(),
+                    requester: PackageName::unchecked("unknown"), // TODO: thread requester
```
What is this todo about?

markusz replied:
Ah, I remember. 

I wanted to have a error message that reads contains the dependency name. Now, all we get is "local dependency foo conflicts with remote dependency required by unknown"

However, I was not very fond of carrying the requesting parent package name through the entire call chain only to be able to log it in that particular error message. 

I solved this by added the offending package to the miette trace, keeping code clean and message meaningful for users

mara-schulke commented on src/resolver.rs:
```diff
@@ -1,469 +1,478 @@
-    type IntoIter = std::collections::hash_map::IntoValues<PackageName, ResolvedDependency>;
+/// Errors that can occur during dependency resolution
+#[derive(Error, Diagnostic, Debug)]
+pub enum DependencyError {
```
Very nice:)

mara-schulke commented on src/workspace.rs:
```diff
@@ -0,0 +1,117 @@
+
+        // Process each member pattern
+        for pattern in member_patterns {
+            if pattern.contains(['*', '?', '[']) {
```
Why not always use glob? Glob will also match static strings!

mara-schulke commented on src/workspace.rs:
```diff
@@ -0,0 +1,117 @@
+                        resolved_members.insert(PathBuf::from(name));
+                    }
+                }
+            } else {
```
It seems that this can be skipped if you always glob

Commit:
fix(cli): Various remaining nits

Commit:
fix(cli): Use miette error wrapping instead of passing parameters through call cain

Commit:
Merge branch 'mz/feat/add-workspace-support' into mz/add-workspace-support/redesign-manifest-impl

# Conflicts:
#	src/command.rs
#	src/manifest.rs
#	src/operations/publisher.rs

Commit:
chore(cli): Post merge cleanup

markusz committed:
Merge pull request #5 from markusz/mz/add-workspace-support/redesign-manifest-impl

Facilitate type system for cleaner manifest handling

Commit:
ci(cli): Fix clippy findings

Commit:
ci(cli): Fix versions flagged in cargo deny check

Commit:
ci(cli): Fix typos flagged in CI

Commit:
ci(cli): Fix unnecessary references

Commit:
ci(cli): Convert let_chains to nested if to account for stable version in nix

Commit:
Revert "ci(cli): Convert let_chains to nested if to account for stable version in nix"

This reverts commit 2575d3b2d6dd28fb8d95d92f989c0634350b7cf1.

Commit:
ci(cli): Updates lockfiles for nix (stable=1.91, supports let chains=) and cargo (slab@0.4.11 to address RUSTSEC-2025-0047)
